### PR TITLE
Accessibility

### PR DESCRIPTION
**Pull request by licarijd**
### Updates

- tabindex each chart data element
- aria label each chart data element as "Label: [x value], Value: [yval], belongs to the dataset [dataset name]"
- mark every chart container as a live region with role=alert so users get notified when charts update

### Review checklist (for reviewers only)

- [ ] Demos all features
- [ ] Documented/annotated
- [ ] Matches UI/UX specs
- [ ] Meets the [code style guide](https://github.com/IBM/carbon-charts/wiki/Code-style-guide)
- [ ] Accessible
- [ ] Mobile first (responsive)
- [ ] RTL support (bidirectional text)
- [ ] Performant (limited bloat)

### CODE REVIEW

File: packages/core/src/bar-chart.ts [hunk 1]
```diff
@@ -106,7 +106,11 @@ export class BarChart extends BaseAxisChart {
 						.attr("fill", d => this.getFillScale()[d.datasetLabel](d.label))
 						.attr("stroke", d => this.options.accessibility ? this.colorScale[d.datasetLabel](d.label) : null)
 						.attr("stroke-width", Configuration.bars.default.strokeWidth)
-						.attr("stroke-opacity", d => this.options.accessibility ? 1 : 0);
+						.attr("stroke-opacity", d => this.options.accessibility ? 1 : 0)
+						// Make the data element accessible and pass x and y values for the data point, and the dataset label
+						.call(this.makeAccessible,
+							(d, index) => `Label: ${d.label}, Value: ${d.value}, belongs to ${d.datasetLabel}`,
+							d => this.y(Math.max(0, d.value)));
```
Comment (theiliad) [inline packages/core/src/bar-chart.ts]: the above 3 lines are very hard to read/understand. Only `makeAccessible` should be called here, and `makeAccessible` should know what to do with the data passed in.

If separate logic is needed for `bar-chart`, then `makeAccessible` should be overriden in `bar-chart`.

File: packages/core/src/base-chart.ts [hunk 2]
```diff
@@ -394,6 +394,8 @@ export class BaseChart {
 			chartId = this.id;
 			container = parent.append("div");
 			container.attr("chart-id", chartId)
+				// Alert user when the contents of the chart container changes
+				.attr("role", "alert")
```
Comment (theiliad) [inline packages/core/src/base-chart.ts]: Not sure if this is the greatest choice. If multiple charts are on the page, then the user could potentially be receiving too many alerts. @zvonimirfras @cal-smith what do you think?

File: packages/core/src/base-chart.ts [hunk 3]
```diff
@@ -951,4 +953,14 @@ export class BaseChart {
 	getBBox(selector: any) {
 		return this.innerWrap.select(selector).node().getBBox();
 	}
+
+	// ================================================================================
+	// Makes chart data components more screen reader friendly
+	// ================================================================================
+	makeAccessible (innerWrap: any, label?: any) {
+		// Make chart data components tabbable
+		innerWrap.attr("tabindex", 0)
```
Comment (theiliad) [inline packages/core/src/base-chart.ts]: innerWrap should not be passed into `makeAccesible`. Can be accessed using `this.innerWrap`.

File: packages/core/src/base-chart.ts [hunk 4]
```diff
@@ -951,4 +953,14 @@ export class BaseChart {
...
+		// Aria label contains the x value, y value, and dataset for the data element
+		.attr("aria-label", label);
```
Comment (theiliad) [inline packages/core/src/base-chart.ts]: Same as my comment above, label should be generated in here.

File: packages/core/src/line-chart.ts [hunk 5]
```diff
@@ -89,7 +89,10 @@ export class LineChart extends BaseAxisChart {
 				.attr("cx", d => this.x(d.label) + margins.left)
 				.attr("cy", d => this.y(d.value))
 				.attr("r", Configuration.charts.pointCircles.radius)
-				.attr("stroke", d => this.colorScale[d.datasetLabel](d.label));
+				.attr("stroke", d => this.colorScale[d.datasetLabel](d.label))
+				.call(this.makeAccessible, d => "Label: " + d.label
+						+ ", Value: " + d.value
+						+ ", belongs to " + d.datasetLabel, d => this.y(Math.max(0, d.value)));
```
Comment (theiliad) [inline packages/core/src/line-chart.ts]: Still using string concatenation here.

File: packages/core/src/bar-chart.ts (hunk 1 above)
Comment (licarijd) [reply packages/core/src/bar-chart.ts]: Just passing a string containing the label, value, and dataset into makeAccessible Separate logic isn't needed for bar chart, every chart will also be passing in a string with those values.

Comment (licarijd) [reply packages/core/src/bar-chart.ts]: Nevermind, I understand after seeing your comment below

Comment (theiliad) [reply packages/core/src/bar-chart.ts]: Ok great. This function is still hard to read though. If you just call `this.makeAccesible`, `d` will be passed into it, and you can do all of this in there

Comment (theiliad) [reply packages/core/src/bar-chart.ts]: ok

File: packages/core/src/base-chart.ts (hunk 3 above)
Comment (licarijd) [reply packages/core/src/base-chart.ts]: I don't have to pass innerWrap in, but leaving innerWrap?:any as a parameter does the trick

File: packages/core/src/base-chart.ts [hunk 6]
```diff
@@ -951,4 +954,28 @@ export class BaseChart {
...
+
+	// ================================================================================
+	// Makes chart data components more screen reader friendly
+	// ================================================================================
+	makeAccessible (element?: any, type?: any, dataList?: any) {
+		// Make chart data components tabbable
+		element.attr("tabindex", 0);
+		// Aria label contains the x value, y value, and dataset for the data element
+		if (type === "axis") {
+			element.attr("aria-label", (d) => `Label: ${d.label}, Value: ${d.value}, belongs to ${d.datasetLabel}`);
```
Comment (zvonimirfras) [inline packages/core/src/base-chart.ts]: How do we translate these strings?

Comment (zvonimirfras) [inline packages/core/src/base-chart.ts]: We don't use optional parenthesis in arrow functions.

File: packages/core/src/base-chart.ts (hunk 2 above)
Comment (cal-smith) [reply packages/core/src/base-chart.ts]: [based on MDN](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alert_role) this definitely wouldn't be a good use of `role="alert"`

Comment (theiliad) [reply packages/core/src/base-chart.ts]: Great thanks Cal

**Pull request closed without merging**